src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -271,5 +274,46 @@ internal static MetadataPropertyName GetMetadataPropertyName(ReadOnlySpan<byte>
 
             return MetadataPropertyName.NoMetadata;
         }
+
+        internal static bool TryGetReferenceFromJsonElement(
+            ref ReadStack state,
+            JsonElement element,
+            out object? referenceValue)
+        {
+            bool refMetadataFound = false;
+            referenceValue = default;
+
+            if (element.ValueKind == JsonValueKind.Object)
+            {
+                int propertyCount = 0;
+                foreach (JsonProperty property in element.EnumerateObject())
+                {
+                    propertyCount++;
+                    if (refMetadataFound)
+                    {
+                        // There are more properties in an object with $ref.
+                        ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties();
+                    }
+                    else if (property.EscapedNameEquals(s_refPropertyName))
```

Comment:
Looks like we lose the validation that the metadata in the payload was the unescaped literal "$ref". Do we care?

Reply:
Hmmm not really, unlike `NameEquals`; `EscapedNameEquals` will compare against the raw bytes, i.e. it won't unescape the name.